File: ios/Reguerta/Reguerta/Domain/Commitments/SeasonalCommitmentRepository.swift
import Foundation

protocol SeasonalCommitmentRepository: Sendable {
    func activeCommitments(userId: String, environment: SessionEnvironment) async throws -> [SeasonalCommitment]
}

/// Resolves commitments through the canonical member-document id.
///
/// The stored-data inventory and shared Firestore contract both require `userId == member.id`. A successful empty
/// result therefore means that the member has no active seasonal commitments; it must not trigger identity fallbacks.
func loadActiveCommitments(
    for member: Member,
    using loadForUser: @Sendable (String) async throws -> [SeasonalCommitment]
) async throws -> [SeasonalCommitment] {
    guard let canonicalKey = member.seasonalCommitmentLookupKeys.first else { return [] }
    return try await loadForUser(canonicalKey).normalizedForMemberLookup
}

private extension Array where Element == SeasonalCommitment {
    var normalizedForMemberLookup: [SeasonalCommitment] {
        Dictionary(
            map { ($0.id, $0) },
            uniquingKeysWith: { _, latest in latest }
        ).values.sorted { lhs, rhs in
            if lhs.seasonKey.localizedCaseInsensitiveCompare(rhs.seasonKey) != .orderedSame {
                return lhs.seasonKey.localizedCaseInsensitiveCompare(rhs.seasonKey) == .orderedAscending
            }
            return lhs.productId.localizedCaseInsensitiveCompare(rhs.productId) == .orderedAscending
        }
    }
}

File: android/Reguerta/app/src/main/java/com/reguerta/user/domain/commitments/SeasonalCommitmentRepository.kt
package com.reguerta.user.domain.commitments

import com.reguerta.user.domain.access.Member

interface SeasonalCommitmentRepository {
    suspend fun getActiveCommitmentsForUser(userId: String): List<SeasonalCommitment>
}

internal fun Member.seasonalCommitmentLookupKeys(): List<String> =
    listOf(id.trim()).filter(String::isNotBlank)

/**
 * Resolves commitments through the canonical member-document id.
 *
 * The stored-data inventory and shared Firestore contract both require `userId == member.id`. A successful empty
 * result therefore means that the member has no active seasonal commitments; it must not trigger identity fallbacks.
 */
internal suspend fun loadActiveCommitmentsForMember(
    member: Member,
    loadForUser: suspend (String) -> List<SeasonalCommitment>,
): List<SeasonalCommitment> {
    val canonicalKey = member.seasonalCommitmentLookupKeys().firstOrNull() ?: return emptyList()
    return loadForUser(canonicalKey).normalizedForMemberLookup()
}

private fun List<SeasonalCommitment>.normalizedForMemberLookup(): List<SeasonalCommitment> =
    distinctBy(SeasonalCommitment::id)
        .sortedWith(compareBy<SeasonalCommitment> { it.seasonKey }.thenBy { it.productId })
